include/votca/tools/matrix.h

```diff
@@ -33,283 +36,301 @@ namespace tools {
+   *
+   * Euler angles are not good for creating random rotations. This function
+   * uses a method proposed by Arvo in Graphics Gems to produce uniform
+   * random rotations.
+   */
+  void RandomRotation();
+
+  /**
+   * \brief calculate eigenvalues and eigenvectors
+   * @param out struct containing eigenvals + eigenvecs
+   */
+  void SolveEigensystem(eigensystem_t &out);
+
+  /**
+   * \brief transpose the matrix
+   * @return the matrix after transpose
+   *
+   * After this operation, matrix stores the transposed value.
+   */ 
+  matrix &Transpose() {
+    std::swap(_m[1], _m[3]);
+    std::swap(_m[2], _m[6]);
+    std::swap(_m[5], _m[7]);
+    return *this;
   }
 
-  inline vec operator*(const matrix &r, const vec &a) { return matrix(r) * a; }
+  /**
+   * \brief matrix-matrix product
+   * @param a the matrix to multiply with
+   * @return multiplied matrix
+   */
+  matrix operator*(const matrix &a) {
+    matrix r;
+    r._m[0] = _m[0] * a._m[0] + _m[1] * a._m[3] + _m[2] * a._m[6];
+    r._m[1] = _m[0] * a._m[1] + _m[1] * a._m[4] + _m[2] * a._m[7];
+    r._m[2] = _m[0] * a._m[2] + _m[1] * a._m[5] + _m[2] * a._m[8];
 
-  inline matrix operator/(const matrix &r, const double &d) {
-    return (matrix(r) /= d);
+    r._m[3] = _m[3] * a._m[0] + _m[4] * a._m[3] + _m[5] * a._m[6];
+    r._m[4] = _m[3] * a._m[1] + _m[4] * a._m[4] + _m[5] * a._m[7];
+    r._m[5] = _m[3] * a._m[2] + _m[4] * a._m[5] + _m[5] * a._m[8];
+
+    r._m[6] = _m[6] * a._m[0] + _m[7] * a._m[3] + _m[8] * a._m[6];
+    r._m[7] = _m[6] * a._m[1] + _m[7] * a._m[4] + _m[8] * a._m[7];
+    r._m[8] = _m[6] * a._m[2] + _m[7] * a._m[5] + _m[8] * a._m[8];
+
+    return r;
   }
-  inline matrix operator+(const matrix &r, const matrix &v) {
-    return (matrix(r) += v);
+
+  /**
+   * \brief matrix-vector product A*x
+   * @param a vector
+   * @return A*x
+   */
+  vec operator*(const vec &a) {
+    return vec(_m[0] * a.getX() + _m[1] * a.getY() + _m[2] * a.getZ(),
+               _m[3] * a.getX() + _m[4] * a.getY() + _m[5] * a.getZ(),
+               _m[6] * a.getX() + _m[7] * a.getY() + _m[8] * a.getZ());
   }
-  inline matrix operator-(const matrix &r, const matrix &v) {
-    return (matrix(r) -= v);
+
+  friend matrix operator*(const double &, const matrix &);
+  friend vec operator*(const vec &, const matrix &);
+
+  template <class Archive>
+  void serialize(Archive &arch, const unsigned int version) {
+    arch &_m;
   }
 
-  /* provided the matrix a diagonalizes it and returns the eigenvalues
-     lambda0 = a[0][0], lambda1 = a[1][1], lambda2= a[2][2], ...
-     as well as the corresponding eigenvectors v[][0], v[][1], v[][2]
-   */
-  int cjcbi(matrix &a, matrix &v, double eps = 1e-10, int jt = 100);
+ private:
+  double _m[9];
+  void checkBounds_(const byte_t row_or_col) const;
+};
+
+inline void matrix::checkBounds_(const byte_t row_or_col) const {
+  assert( row_or_col > boost::lexical_cast<byte_t>(-1) && row_or_col < boost::lexical_cast<byte_t>(3));
```

Comment:
why byte? Additionally the assert should have a message E.g. "Access beyond array boundary"

Reply:
Also why lexical_cast, and why -1 with byte which is unsigned

Reply:
I guess there is no need to check the lower bound than. Thanks for pointing that out.